docs/en/lfs.md

```diff
@@ -1,205 +1,144 @@
-    return ba and "Module not in LFS" or fn
-  end
-  table.insert(package.loaders, loader_flash)
-
-end
-```
+See [lua_examples/lfs/_init.lua](https://github.com/nodemcu/nodemcu-firmware/tree/dev/lua_examples/lfs/_init.lua) for the code that I use in my `_init` module to do create a simple access API for LFS.  There are two parts to this.
 
-This code is also conveniently packaged in [lua_examples/lfs/_init.lua](../../lua_examples/lfs/_init.lua).
+The first sets up an LFS table in `_G` with the `__index` and `__newindex` metamethods. The main purpose of the `__index()` is to resolve any names against the LFS using a `node.flashindex()` call, so that `LFS.someFunc(params)` does exactly what you would expect it to do: this will call `someFunc` with the specified parameters, if it exists in in the LFS.  The LFS properties `_time`, `_config` and `_list` can be used to access the other LFS metadata that you need.  See the code to understand what they do, but `LFS._list` is the array of all module names in the LFS.
 
-Once this has been executed, if you have a function module `func1` in LFS, then `LFS.func1(x,y,z)` just works as you would expect.  The LFS properties `_time`, `_config` and `_list` can be used to access the other LFS metadata that you need.
+The second uses standard Lua functionality to add the LFS to the require [package.loaders](http://pgl.yoyo.org/luai/i/package.loaders) list.  There are four standard loaders, which the require loader searches in turn.  However, NodeMCU only uses the second of these (the Lua loader from the file system).  Since loaders 1,3 and 4 aren't used, we can simply replace the 1st or the 3rd by code to use `node.flashindex()` to return the LFS module.  The supplied `_init` puts the LFS loader at entry [3], so if the module is in both SPIFFS and LFS, then the SPIFFS version will be loaded, so if you have an out of date version in SPIFFS then it will still get loaded instead of the one if LFS (and I have been burnt by this.)
 
-Of course, if you use Lua modules to build your application then `require "some_module"` will automatically path in and load your modules from LFS.  Note that SPIFFS is still ahead of LFS in this search list, so if you have a dev version in SPIFFS, say, then this will be loaded first. However, if you want to want to swap this search order so that the LFS is searched first, then run `table.insert(package.loaders,2,loader_flash)` in your `_init` code.  If you need to swap the search order temporarily for development or debugging, then do this after you've run the `_init` code:
+If you want to swap this search order so that the LFS is searched first, then SET `package.loaders[1] = loader_flash` in your `_init` code.  If you need to swap the search order temporarily for development or debugging, then do this after you've run the `_init` code:
 
 ```Lua
-do local pl = package.loaders; pl[2],pl[5] = pl[5],pl[2]; end
+do local pl = package.loaders; pl[1],pl[3] = pl[3],pl[1]; end
 ```
```

Comment:
I remain somewhat unhappy about exposing these particular details to programmers, but for the moment it's fine, I suppose.

Reply:
To quote the LFS paper: "The second uses standard Lua functionality to add the LFS to the require [package.loaders](http://pgl.yoyo.org/luai/i/package.loaders) list".  This is already openly documented, but we can refine it later.

Reply:
I mean, I don't think we should be hiding that `package.loaders[1]` is an interesting thing, I just would wish for a way to not encourage that particular thing to end up in people's lua, since we might change `package.loaders` in the future.  But as you say, let's worry about refining it later, not now.